dynamic-config/config-apollo/src/main/java/io/servicecomb/config/client/ApolloClient.java

```diff
@@ -0,0 +1,162 @@
+
+public class ApolloClient {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ApolloClient.class);
+
+  private static final ApolloConfig APOLLO_CONFIG = ApolloConfig.INSTANCE;
+
+  public static final Map<String, Object> originalConfigMap = new HashMap<>();
+
+  private static ScheduledExecutorService EXECUTOR = Executors.newScheduledThreadPool(1);
+
+  private int refreshInterval = APOLLO_CONFIG.getRefreshInterval();
+
+  private int firstRefreshInterval = APOLLO_CONFIG.getFirstRefreshInterval();
+
+  private String serviceUri = APOLLO_CONFIG.getServerUri();
+
+  private String serviceName = APOLLO_CONFIG.getServiceName();
+
+  private String token = APOLLO_CONFIG.getToken();
+
+  private String env = APOLLO_CONFIG.getEnv();
+
+  private String clusters = APOLLO_CONFIG.getServerClusters();
+
+  private String namespace = APOLLO_CONFIG.getNamespace();
+
+  private UpdateHandler updateHandler;
+
+  private static ClientPoolManager<HttpClientWithContext> clientMgr = new ClientPoolManager<>();
+
+  public ApolloClient(UpdateHandler updateHandler) {
+    this.updateHandler = updateHandler;
+  }
+
+  public void refreshApolloConfig() {
+    EXECUTOR
+        .scheduleWithFixedDelay(new ConfigRefresh(serviceUri), firstRefreshInterval, refreshInterval, TimeUnit.SECONDS);
+  }
+
+  class ConfigRefresh implements Runnable {
+    private String serviceUri;
+
+    public ConfigRefresh(String serviceUris) {
+      this.serviceUri = serviceUris;
+    }
+
+    @Override
+    public void run() {
+      try {
+        refreshConfig();
+      } catch (Exception e) {
+        LOGGER.error("client refresh thread exception", e);
+      }
+    }
+
+    public void refreshConfig() {
+      RestTemplate rest = new RestTemplate();
+      HttpHeaders headers = new HttpHeaders();
+      headers.add("Content-Type", "application/json;charset=UTF-8");
+      headers.add("Authorization", token);
+      HttpEntity<String> entity = new HttpEntity<String>(headers);
+      ResponseEntity<String> exchange = rest.exchange(composeAPI(), HttpMethod.GET, entity, String.class);
+      if (HttpResponseStatus.OK.code() == exchange.getStatusCode().value()) {
+        try {
+          Map<String, Object> body = JsonUtils.OBJ_MAPPER.readValue(exchange.getBody(),
+              new TypeReference<Map<String, Object>>() {
+              });
+          refreshConfigItems((Map<String, Object>) body.get("configurations"));
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+
+    private String composeAPI() {
+      String api = serviceUri + "/openapi/v1/envs/" + env +
+          "/apps/" + serviceName +
+          "/clusters/" + clusters +
+          "/namespaces/" + namespace +
+          "/releases/latest";
+      return api;
+    }
+
+    private void refreshConfigItems(Map<String, Object> map) {
+      compareChangedConfig(originalConfigMap, map);
+      originalConfigMap.clear();
```

Comment:
originalConfigMap is a public static variable, and it's a HashMap
so is the clear and putAll safe?

Reply:
Use public static variable would be better.

Reply:
Use ConcurrentHashMap instead.